### PR TITLE
Feature/fileapi pdf protector

### PR DESCRIPTION
**Pull request by FXKewang**
Foxit has finished development for PDF Protector in MIP SDK and has written tests now using GoogleTest and GoogleMock. It is being built along with the fileapi_ut build in the branch Feature/fileapi-pdf-protector.
Now the PDF protector passes the gTest and Foxit creates a pull request to merge to ‘fileapi’ branch.

### CODE REVIEW

File: samples/rms_sample/mainwindow.cpp [hunk 1]
```diff
@@ -324,7 +324,7 @@ void MainWindow::FileAPIEncryptRights(const string &fileIn,
     auto inFile = make_shared<fstream>(
       fileIn, ios_base::in | ios_base::out | ios_base::binary);
 
-    std::string filename = fileIn.substr(fileIn.find_last_of("\\/") + 1);
+    std::string filename = fileIn;//fileIn.substr(fileIn.find_last_of("\\/") + 1);
```
Comment (vishmittal) [inline samples/rms_sample/mainwindow.cpp:326]: Why is this needed?

File: sdk/rms_sdk/FileAPI/PDFProtector.h [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+#include "UserPolicy.h"
+#include "PDFObjectModel.h"
+#include "CryptoAPI.h"
+
+using namespace rmscore::pdfobjectmodel;
+
+namespace rmscore {
+namespace fileapi {
+
+#define MIN_RAW_SIZE 64 * 1024 * 1024
```
Comment (vishmittal) [inline sdk/rms_sdk/FileAPI/PDFProtector.h]: add a comment to explain what is this

File: samples/rms_sample/mainwindow.cpp [hunk 3]
```diff
@@ -871,3 +871,182 @@ void MainWindow::AddLog(const QString& tag, const QString& message) {
...
+    string fileIn = SelectFile("Select a PDF file to decrypt");
+
+    if (!fileIn.empty()) {
+      PDFFileDecrypt(
+        fileIn,
+        ui->lineEdit_clientId->text().toStdString(),
+        ui->lineEdit_redirectUrl->text().toStdString(),
+        ui->lineEdit_clientEmail->text().toStdString());
+    }
+}
```
Comment (vishmittal) [inline samples/rms_sample/mainwindow.cpp:1051]: update the sample app to protect pdf using custom rights as well.

File: sdk/rms_sdk/FileAPI/PDFProtector.cpp [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+
+namespace logger = rmscore::platform::logger;
+
+namespace rmscore {
+namespace fileapi {
+
+////////////////////////////////////////////////////////////////////////////
+/// class PDFProtector
+PDFCryptoHandlerImpl::PDFCryptoHandlerImpl(std::shared_ptr<PDFProtector> pdf_protector)
+{
```
Comment (vishmittal) [inline sdk/rms_sdk/FileAPI/PDFProtector.cpp]: Please follow the coding guidelines and have the opening braces on the prev line. Here and all other places for all methods, loops, if-else etc.

`PDFCryptoHandlerImpl::PDFCryptoHandlerImpl(std::shared_ptr<PDFProtector> pdf_protector) {`

[http://google.github.io/styleguide/cppguide.html#Formatting](url)

File: sdk/rms_sdk/FileAPI/ProtectorSelector.h [hunk 5]
```diff
@@ -24,16 +26,35 @@ enum class ProtectorType {
     PDF = 4,
 };
 
-class ProtectorSelector
+class IProtectorSelector
```
Comment (vishmittal) [inline sdk/rms_sdk/FileAPI/ProtectorSelector.h:29]: Do we need an interface here? If not, let's remove it

File: sdk/rms_sdk/PDFObjectModel/Basic.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  virtual FX_BOOL IsEOF();
+
+  virtual FX_FILESIZE GetPosition();
+
+  virtual FX_BOOL ReadBlock(void* buffer, FX_FILESIZE offset, size_t size);
+
+  virtual size_t ReadBlock(void* buffer, size_t size);
+
+  virtual	FX_BOOL WriteBlock(const void* buffer, FX_FILESIZE offset, size_t size);
+
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/Basic.h]: Nit: remove empty space

File: sdk/rms_sdk/PDFObjectModel/Basic.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+typedef unsigned long   UTF32;  /* at least 32 bits */
+typedef unsigned short  UTF16;  /* at least 16 bits */
+typedef unsigned char   UTF8;   /* typically 8 bits */
+typedef unsigned int    INT;
+
+/**
+ * @brief It is the utility class for pdf object model, like encoding converting.
+ */
+
+namespace utility {
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/Basic.h]: please change this namespace to 'PDFUtils'

File: sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+   * @param[in] size              The input buffer size.
+   * @return true for success, otherwise false.
+   */
+  virtual bool AppendBlock(const void* buffer_pointer, uint32_t size) = 0;
+};
+
+#define  PDFWRAPPERDOC_TYPE_UNKNOWN -1	/** Unknown. */
+#define  PDFWRAPPERDOC_TYPE_NORMAL	0	/** Normal document. */
+#define  PDFWRAPPERDOC_TYPE_IRMV1	1	/** For IRM V1 wrapper document. */
+#define  PDFWRAPPERDOC_TYPE_IRMV2	2	/** For IRM V2 wrapper document. */
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h]: you could have an enum instead of macros

File: sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+   */
+  virtual uint32_t GetWrapperType() const = 0;
+
+  /**
+   * @brief Gets the cryptographic filter name and version.
+   * @param[out] graphic_filter  It receives the cryptographic filter name.
+   * @param[out] version_num         It receives the cryptographic filter version.
+   * @return true for success, otherwise false.
+   */
+  virtual bool GetCryptographicFilter(std::wstring& graphic_filter, float &version_num) const = 0;
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h]: Avoid usage of wstring at all throughout the change since it's size is platform dependent.

File: sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+   * @brief The PDF protector checks the publishing license validity here. If false
+   * is returned, UnprotectCustomEncryptedFile will fail.
+   */
+  virtual bool OnInit(unsigned char* publishing_license,
+                      uint32_t publishing_license_size) = 0;
+
+  virtual std::shared_ptr<PDFCryptoHandler> CreateCryptoHandler() = 0;
+};
+
+#define PDFCREATOR_ERR_SUCCESS				0
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h]: use class enum for these codes

File: sdk/rms_sdk/PDFObjectModel/PDFModuleMgr.h [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+class PDFModuleMgrImpl : public PDFModuleMgr {
+ public:
+  explicit PDFModuleMgrImpl();
+
+  static void RegisterSecurityHandler(const std::string& filter_name,
+                                      std::shared_ptr<PDFSecurityHandler> security_hander);
+  virtual  ~PDFModuleMgrImpl();
+
+ private:
+  CPDF_ModuleMgr* pdf_module_manager_;
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFModuleMgr.h]: use smart pointers for these data members and use custom deleters with the smart pointers to destroy or free memory etc.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.h [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+#define RMSSDK_PDFOBJECTMODEL_PDFCREATOR_H_
+
+#include "PDFObjectModel.h"
+#include "Core.inc"
+
+namespace rmscore {
+namespace pdfobjectmodel {
+
+#define PROGRESSIVE_ENCRYPT_TEMP_FILE L".RMS.PE.temp"
+
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.h]: Don't use wide strings unless abosultely needed.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+  return nullptr;
+}
+
+FX_BOOL CustomCryptoHandler::DecryptStream(
+    FX_LPVOID context,
+    FX_LPCBYTE src_buf,
+    FX_DWORD src_size,
+    CFX_BinaryBuf& dest_buf) {
+  if (pdf_crypto_handler_) {
+    std::shared_ptr<CFX_BinaryBuf> shared_dest_buf(&dest_buf, [=](CFX_BinaryBuf* pDestBuf) {
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: why do you need this?

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 14]
```diff
@@ -0,0 +1,672 @@
...
+     PDFBinaryBufImpl bin_buf(shared_dest_buf);
+     return pdf_crypto_handler_->DecryptStream((char*)src_buf, src_size, &bin_buf);
+  }
+  return false;
+}
+
+FX_BOOL CustomCryptoHandler::DecryptFinish(FX_LPVOID context,
+                                           CFX_BinaryBuf& dest_buf) {
+  if (pdf_crypto_handler_) {
+    std::shared_ptr<CFX_BinaryBuf> shared_dest_buf(&dest_buf, [=](CFX_BinaryBuf* pDestBuf) {
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: Why are you creating a smart pointer from the reference?

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 15]
```diff
@@ -0,0 +1,672 @@
...
+    PDFBinaryBufImpl bin_buf(shared_dest_buf);
+    return pdf_crypto_handler_->ProgressiveEncryptFinish(&bin_buf);
+  }
+  return false;
+}
+
+//////////////////////////////////////////////////////////////////////////
+//class CustomProgressiveEncryptHandler
+
+CustomProgressiveEncryptHandler::CustomProgressiveEncryptHandler(CFX_WideString temp_path) {
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: do we need to use widestring? 
Would files protected on Windows work on Unix and vice-versa?

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 16]
```diff
@@ -0,0 +1,672 @@
...
+
+CustomProgressiveEncryptHandler::~CustomProgressiveEncryptHandler() {
+  crypto_handler_ = nullptr;
+}
+
+
+IFX_FileStream* CustomProgressiveEncryptHandler::GetTempFile() {
+  CFX_WideString path = temp_path_;
+  path += PROGRESSIVE_ENCRYPT_TEMP_FILE;
+  IFX_FileStream*	file_stream = FX_CreateFileStream(path, FX_FILEMODE_Truncate);
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: use smart pointer where memory allocation is taking place.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 17]
```diff
@@ -0,0 +1,672 @@
...
+
+FX_BOOL CustomProgressiveEncryptHandler::UpdateFilter(CPDF_Dictionary* pdf_dictionary) {
+    return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+//class CustomSecurityHandler
+
+CustomSecurityHandler::CustomSecurityHandler(
+    std::shared_ptr<PDFSecurityHandler> pdf_security_handler) {
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: use shared_ptr only if needed. Else use unique_ptr

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 18]
```diff
@@ -0,0 +1,672 @@
...
+  }
+
+  return true;
+}
+
+FX_DWORD CustomSecurityHandler::GetPermissions() {
+  return 0;
+}
+
+FX_BOOL CustomSecurityHandler::IsOwner() {
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: why is this always returning true?

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 19]
```diff
@@ -0,0 +1,672 @@
...
+  return true;
+}
+
+FX_BOOL CustomSecurityHandler::GetCryptInfo(int& cipher, FX_LPCBYTE& buffer, int& keylen) {
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: Why is this always returning true? Is this not implemented?

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 20]
```diff
@@ -0,0 +1,672 @@
...
+  return true;
+}
+
+FX_BOOL CustomSecurityHandler::IsMetadataEncrypted() {
+  return encrypt_metadata_;
+}
+
+CPDF_CryptoHandler* CustomSecurityHandler::CreateCryptoHandler() {
+  CPDF_CryptoHandler* core_pdf_crypto_handler = nullptr;
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: use unique ptr here since you are allocating memory.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp [hunk 21]
```diff
@@ -0,0 +1,672 @@
...
+  }
+
+  return result;
+}
+
+CPDF_Dictionary* PDFCreatorImpl::CreateEncryptionDict(
+    const std::string& filter_name,
+    const std::vector<unsigned char> &publishing_license) {
+  CPDF_Dictionary* encryption_dictionary = new CPDF_Dictionary;
+
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: use smart pointers where memory is being allocated. For example with 'new' keyword.

**Review by vishmittal — CHANGES_REQUESTED** [review]

File: samples/rms_sample/mainwindow.cpp (hunk 1 above)
Comment (FXKewang) [reply samples/rms_sample/mainwindow.cpp:326]: Because we have to create a temp file to storage the large cache data under the folder of the input original file when there are some large PDF objects like video. Do you have any suggestion?

File: sdk/rms_sdk/FileAPI/ProtectorSelector.h (hunk 5 above)
Comment (FXKewang) [reply sdk/rms_sdk/FileAPI/ProtectorSelector.h:29]: Yes. We need it. Otherwise the existing unit test of FileAPIProtectorSelectorTest.cpp cannot be linked correctly.

File: sdk/rms_sdk/PDFObjectModel/Basic.h (hunk 7 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/Basic.h]: Are you sure? According to Google C++ Style, namespace names are all lower-case.

**Comment by slreznit** [comment]
Side question - I couldn't find the code that reads/writes PDF doc metadata. Is it planned to add this functionality? We will surely need it.

**Comment by FXKewang** [comment]
@slreznit  If you want to reads/writes PDF doc metadata, you can get XML data of metadata by following codes.
CPDF_Dictionary *pRoot = pDoc->GetRoot();
CPDF_Stream *pStream = pRoot->GetStream(FX_BSTRC("Metadata"));
CPDF_StreamAcc acc;
acc.LoadAllData(pStream, FALSE);
int size = acc.GetSize();
FX_LPCBYTE pBuf = acc.GetData();

File: samples/rms_sample/mainwindow.cpp (hunk 3 above)
Comment (FXKewang) [reply samples/rms_sample/mainwindow.cpp:1051]: Done. Git 795b8d8..a4550a4

File: sdk/rms_sdk/FileAPI/PDFProtector.h (hunk 2 above)
Comment (FXKewang) [reply sdk/rms_sdk/FileAPI/PDFProtector.h]: Done. Git a4550a4..68bb8a2

File: sdk/rms_sdk/PDFObjectModel/Basic.h (hunk 6 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/Basic.h]: Done. Git 68bb8a2..2750f5c

File: sdk/rms_sdk/FileAPI/PDFProtector.cpp (hunk 4 above)
Comment (FXKewang) [reply sdk/rms_sdk/FileAPI/PDFProtector.cpp]: Done on 30th Oct. Git 85ed470..c0bd745

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 13 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: At the beginning, I modified the raw pointer to smart pointer as much as possible to meet the requirement. But smart pointer is not good for performance. So it is not needed here. I have modified it. Git 2750f5c..d4e13b1

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 14 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: At the beginning, I modified the raw pointer to smart pointer as much as possible to meet the requirement. But smart pointer is not good for performance. So it is not needed here. I have modified it. Git  2750f5c .. d4e13b1

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 15 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: Use widestring to deal with wide character. Yes, files protected on Windows work on Unix and vice-versa.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 16 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: CustomProgressiveEncryptHandler is inherited from foxit core interface. Using smart pointer here is not good for performance.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 17 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: Yes. shared_ptr is needed here.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 18 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: PDF Object Model is for PDF protection and unprotection. The upper layer RMS system controls the permission.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 19 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: This interface is for PDF standard encryption\decryption algorithm. RMS has its own algorithm to deal with the data in CustomCryptoHandler.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 20 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: This interface is inherited from foxit core interface. Foxit pdf core takes over core_pdf_crypto_handler. Using smart pointer here is not good for performance.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp (hunk 21 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp]: This interface is inherited from foxit core interface. Foxit pdf core takes over core_pdf_crypto_handler. Using smart pointer here is not good for performance.

File: sdk/rms_sdk/PDFObjectModel/PDFCreator.h (hunk 12 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFCreator.h]: Use widestring to deal with wide character. Yes, files protected on Windows work on Unix and vice-versa.

File: sdk/rms_sdk/PDFObjectModel/PDFModuleMgr.h (hunk 11 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFModuleMgr.h]: It is foxit core interface and foxit core encapsulate the Create and Release interfaces. It is not suitable to use smart pointers here.

File: sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h (hunk 8 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h]: Done. Git d4e13b1..db2a307

File: sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h (hunk 10 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h]: Done. Git d4e13b1..db2a307

File: sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h (hunk 9 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/PDFObjectModel.h]: widestring is needed in foxit pdf core to deal with wide character.

**Comment by vishmittal** [comment]
>     }

use CBC4K here

---
Refers to: samples/rms_sample/mainwindow.cpp:63 in 44c04c3. [](commit_id = 44c04c3bbff6577878bc5290d93078a21b04f994, deletion_comment = False)

Kewang: Done!

**Comment by vishmittal** [comment]
>         rmscore::fileapi::ProtectWithCustomRightsOptions pt (rmscore::fileapi::CryptoOptions::AES128_ECB,

Use CBC4K here

---
Refers to: samples/rms_sample/mainwindow.cpp:1094 in 44c04c3. [](commit_id = 44c04c3bbff6577878bc5290d93078a21b04f994, deletion_comment = False)

Kewang: Done!

**Comment by vishmittal** [comment]
@ryanbest is added to the review. #Closed

File: sdk/rms_sdk/PDFObjectModel/pdf_module_mgr.cpp [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+std::shared_ptr<PDFSecurityHandler> PDFModuleMgrImpl::GetSharedSecurityHandler() {
+  return shared_security_hander_;
+}
+
+static CPDF_SecurityHandler* CreateCustomerSecurityHandler(void* param) {
+  FX_UNREFERENCED_PARAMETER(param);
+  PDFModuleMgrImpl* pdf_module_mgr = (PDFModuleMgrImpl*)g_pdf_module_mgr.get();
+  std::shared_ptr<PDFSecurityHandler> security_hander = pdf_module_mgr->GetSharedSecurityHandler();
+  //core takes over custom_security_handler
+  CustomSecurityHandler* custom_security_handler = new CustomSecurityHandler(security_hander);
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/pdf_module_mgr.cpp:63]: Where is this memory deallocated?

File: sdk/rms_sdk/PDFObjectModel/pdf_creator.cpp [hunk 23]
```diff
@@ -0,0 +1,679 @@
...
+CPDF_Dictionary* PDFCreatorImpl::CreateEncryptionDict(
+    const std::string& filter_name,
+    const std::vector<unsigned char> &publishing_license) {
+  CPDF_Dictionary* encryption_dictionary = new CPDF_Dictionary;
+
+  encryption_dictionary->SetAtName("Filter", filter_name.c_str());
+
+  CPDF_Dictionary* pCFDic = new CPDF_Dictionary;
+
+  CPDF_Dictionary* crypt_filter_dictionary = new CPDF_Dictionary;
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/pdf_creator.cpp:578]: Where is this memory deallocated?

File: sdk/rms_sdk/PDFObjectModel/pdf_creator.cpp [hunk 24]
```diff
@@ -0,0 +1,679 @@
...
+}
+
+CPDF_Dictionary* PDFCreatorImpl::CreateEncryptionDict(
+    const std::string& filter_name,
+    const std::vector<unsigned char> &publishing_license) {
+  CPDF_Dictionary* encryption_dictionary = new CPDF_Dictionary;
+
+  encryption_dictionary->SetAtName("Filter", filter_name.c_str());
+
+  CPDF_Dictionary* pCFDic = new CPDF_Dictionary;
```
Comment (vishmittal) [inline sdk/rms_sdk/PDFObjectModel/pdf_creator.cpp:576]: Where is this memory deallocated?

Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/pdf_creator.cpp:576]: pCFDic is added to encryption_dictionary. pCFDic will be released in encryption_dictionary's destructor function CPDF_Dictionary::~CPDF_Dictionary(). encryption_dictionary will be released in CPDF_Parser::~CPDF_Parser().

File: sdk/rms_sdk/PDFObjectModel/pdf_creator.cpp (hunk 23 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/pdf_creator.cpp:578]: crypt_filter_dictionary is added to pCFDic. crypt_filter_dictionary will be released in pCFDic's destructor function. pCFDic is added to encryption_dictionary. pCFDic will be released in encryption_dictionary's destructor function CPDF_Dictionary::~CPDF_Dictionary(). encryption_dictionary will be released in CPDF_Parser::~CPDF_Parser().

File: sdk/rms_sdk/PDFObjectModel/pdf_module_mgr.cpp (hunk 22 above)
Comment (FXKewang) [reply sdk/rms_sdk/PDFObjectModel/pdf_module_mgr.cpp:63]: It will be release in function void CPDF_Parser::ReleaseEncryptHandler() when CPDF_Parser is released in CPDF_Parser::~CPDF_Parser().
